nightfall-deployer/contracts/X509.sol

```diff
@@ -0,0 +1,345 @@
+        tlvs = walkDerTree(certificate, 0, tlvLength);
+        // extract the data from the certificate necessary for checking the signature and (hopefully) find the Authority public key in
+        // the smart contract's list of trusted keys
+        bytes32 authorityKeyIdentifier = extractAuthorityKeyIdentifier(tlvs);
+        bytes memory signature = getSignature(tlvs, tlvLength);
+        bytes memory message = getMessage(tlvs);
+        RSAPublicKey memory publicKey = trustedPublicKeys[authorityKeyIdentifier];
+        // validate the cert's signature and check that the cert is in date, record the expiry date against msg.sender
+        checkSignature(signature, message, publicKey);
+        uint256 expiry = checkDates(tlvs);
+        // The certificate is valid and linked to a root we trust, so now we trust the certificate's public key too. Let's add it to our list of trusted keys
+        RSAPublicKey memory certificatePublicKey = extractPublicKey(tlvs);
+        bytes32 subjectKeyIdentifier = extractSubjectKeyIdentifier(tlvs);
+        require(!revokedKeys[subjectKeyIdentifier], 'The key of this certificate has been revoked');
+        // finally, before we can whitelist msg.sender, we should check that they are indeed the owner of the cert (certs are public, after all)
+        // we do that by getting them to sign msg.sender with the private key corresponding to their certificate public key
+        if (!addAddress) {
+            // if we're not adding an address, check that this certificate can sign certificates (because it must be an intermediate one)
+            checkKeyUsage(tlvs, usageBitMaskIntermediate);
+            // if yes, we'll trust it
+            trustedPublicKeys[subjectKeyIdentifier] = certificatePublicKey;
+            return; // we may want to add an intermediate cert to the contract but not add an address.
+        }
+        // as we are trying to add an address, this certificate should be an end user certificate, created for digital signature
+        // and non-repudiation (or possibly other things - we can change this).
+        checkKeyUsage(tlvs, usageBitMaskEndUser);
+        trustedPublicKeys[subjectKeyIdentifier] = certificatePublicKey;
+        checkSignature(
+            addressSignature,
+            abi.encodePacked(uint160(msg.sender)),
+            certificatePublicKey
+        );
+        expires[msg.sender] = expiry;
+        keysByUser[msg.sender] = subjectKeyIdentifier;
+        addUserToWhitelist(msg.sender); // all checks have passed, so they are free to trade for now.
+    }
+
+    // performs an ongoing KYC check (is the user still in the whitelist? Has the public key been revoked? Is the cert in date?)
+    // We could also remove the user in this function, but that would burn more gas.
+    function kycCheck(address user) external view returns (bool) {
+        if (
+            !whitelisting ||
+            (!revokedKeys[keysByUser[user]] &&
```

Comment:
Do you think there should be a way to recover revoked addresses in case we redeploy contracts?

Reply:
Yes, good call.  The list of revoked addresses is public so it can be read anyway, so nothing to do to the smart contract but it's a useful admin function to include.  I'll add it to the admin PR but not here.